buildpacks/go/src/layers/deps.rs

```diff
@@ -1,108 +1,82 @@
-use crate::{cmd, GoBuildpack, GoBuildpackError};
+use crate::{GoBuildpack, GoBuildpackError};
 use libcnb::build::BuildContext;
-use libcnb::data::layer_content_metadata::LayerTypes;
-use libcnb::layer::{ExistingLayerStrategy, Layer, LayerData, LayerResult, LayerResultBuilder};
+use libcnb::data::layer_name;
+use libcnb::layer::{
+    CachedLayerDefinition, InvalidMetadataAction, LayerRef, LayerState, RestoredLayerAction,
+};
 use libcnb::layer_env::{LayerEnv, Scope};
-use libcnb::{Buildpack, Env};
 use libherokubuildpack::log::log_info;
 use serde::{Deserialize, Serialize};
 use std::fs;
-use std::path::Path;
 
-const LAYER_VERSION: &str = "1";
 const MAX_CACHE_USAGE_COUNT: f32 = 100.0;
 const CACHE_ENV: &str = "GOMODCACHE";
 const CACHE_DIR: &str = "cache";
 
-/// A layer that caches the go modules cache
-pub(crate) struct DepsLayer {
-    pub(crate) go_env: Env,
-}
-
 #[derive(Deserialize, Serialize, Clone, PartialEq)]
 pub(crate) struct DepsLayerMetadata {
     // Using float here due to [an issue with lifecycle's handling of integers](https://github.com/buildpacks/lifecycle/issues/884)
     cache_usage_count: f32,
-    layer_version: String,
 }
 
 #[derive(thiserror::Error, Debug)]
 pub(crate) enum DepsLayerError {
     #[error("Couldn't create Go modules cache layer: {0}")]
     Create(std::io::Error),
-    #[error("Couldn't clean Go modules cache: {0}")]
-    Clean(#[from] cmd::Error),
 }
 
-impl Layer for DepsLayer {
-    type Buildpack = GoBuildpack;
-    type Metadata = DepsLayerMetadata;
-
-    fn types(&self) -> LayerTypes {
-        LayerTypes {
+/// Create or restore the layer for the go modules cache (non-vendored dependencies)
+pub(crate) fn handle_deps_layer(
+    context: &BuildContext<GoBuildpack>,
+) -> libcnb::Result<LayerRef<GoBuildpack, (), f32>, GoBuildpackError> {
+    let layer_ref = context.cached_layer(
+        layer_name!("go_deps"),
+        CachedLayerDefinition {
             build: true,
             launch: false,
-            cache: true,
-        }
-    }
+            invalid_metadata_action: &|_| InvalidMetadataAction::DeleteLayer,
+            restored_layer_action: &|restored_metadata: &DepsLayerMetadata, _| {
+                if restored_metadata.cache_usage_count >= MAX_CACHE_USAGE_COUNT {
+                    log_info("Expired Go modules cache");
```

Comment:
(optional) I think Manuel had intended these closures to be side-effect free, and for logging to be performed as part of inspecting `layer_ref.state` later? One benefit of doing that would also be that the `EmptyLayerCause::InvalidMetadataAction` case could be given as message too, eg like:
https://github.com/heroku/buildpacks-python/blob/35c65bd8375547fb6ef00d568df0943015e8a198/src/layers/python.rs#L59-L78
(since at the moment if the metadata is invalid/has changed format, then the cache is silently discarded)